### ISSUE TITLE
Can't require() zeroclipboard

### ISSUE THREAD

**Issue by DelvarWorld** [zeroclipboard/zeroclipboard#568]
According to [this entry](https://github.com/zeroclipboard/zeroclipboard/blob/master/docs/instructions.md#amd) this should work?

    npm install zeroclipboard
    node
    > require('zeroclipboard')
    TypeError: Cannot read property 'currentScript' of undefined
        at _getCurrentScriptUrll/node_modules/zeroclipboard/dist/ZeroClipboard.js:216:18)

Additionally, when I try to require it in the browser with a webpack build, I'm getting `Uncaught Error: Cannot find module "zeroclipboard"`, but I'm not sure if that's related to the same issue or not.

**Comment by JamesMGreene** [comment]
ZeroClipboard is heavily bound to the DOM, so loading it in Node.js doesn't really make any sense (nor is it supported).

Can't speak to WebPack with any experience but we've certainly been around the block with AMD and CommonJS users before (RequireJS, curl.js, node-webmake, Browserify, etc.) but -- AFAIK -- we are still functional in all such environments.

**Comment by DelvarWorld** [comment]
In terms of testing (reactjs) components - we can certainly stub out the component that uses zeroclipboard, but as it's in the dependency tree already I suspect it would break any component tests.

There's probably some boilerplate with jsdom that will make this work (it at least needs document, window.location, and at least one script tag in jsdom), but it would be nice if the script require worked regardless. Maybe even just bail if there's no browser / window?

**Comment by JamesMGreene** [comment]
I see no reason why a non-functional script should just bail (i.e. pretend nothing is wrong)... if anything, I would have it throw an `Error` immediately if such a state is detectable.  However, if you have JSDOM loaded as well, that state will **not** be detectable. :cry:

**Comment by DelvarWorld** [comment]
Yeah, bailing probably isn't ideal. There are modules designed to work in the browser, such as jquery, which you can require without error in node land. The issue is that zeroclipboard reads many things from the document / window **at load time** which doesn't seem ideal to me. That means simply requiring the library has side effects. A solution might be to instantiate all the variables based on `window` etc when you *instantiate* a `new Zeroclipboard()`. If you add syntax to support `require` and publish it to npm, I don't think it's unreasonable for users to think they can require it in nodejs.

If it errors at `require()` time in node land, it makes it impractical to unit test that component, because you can't even require the component itself to stub out dependencies. React components are obviously designed to be run in the browser, but they are also very unit testable in node land, unless some dependencies won't load :(

**Comment by DelvarWorld** [comment]
This did indeed break our unit test suite, as a temporary workaround we're doing something like:

    NODE_ENV=testing testcommand ...

And in your source:

    var ZeroClipboard = process.env.NODE_ENV === "testing" ? function() {} : require("zeroclipboard");

**Comment by JamesMGreene** [comment]
@DelvarWorld: Can you offer up an example of a React component source code that demonstrates how they are making them work well for testing in both the browser and in Node? That would be very helpful. Thanks!

**Comment by DelvarWorld** [comment]
https://facebook.github.io/react/docs/test-utils.html ?

**Comment by DelvarWorld** [comment]
The best friends club:
- https://github.com/ccampbell/mousetrap/issues/283
- https://github.com/zeroclipboard/zeroclipboard/issues/568
- https://github.com/thlorenz/brace/issues/30
